### PR TITLE
Changed form,dashboard

### PR DESCRIPTION
**Pull request by rav-indra**
- Changed form style
-  Added new page to show sold properties
- Changed dashboard style
- Revoke edit and delete access from seller after property sold

### CODE REVIEW

File: app/views/dashboard/_buyer.html.erb [hunk 1]
```diff
@@ -1,15 +1,45 @@
-<h1 style="margin-top: 120px; text-align:center;">Purchased Properties</h1>
 <% props = Property.where(buyer_id: current_user.id) %>
```
Comment (ghost) [inline app/views/dashboard/_buyer.html.erb]: View is only for presenting the data, Not for fetching or collecting the data. So please remove this line from here.

File: app/views/dashboard/_buyer.html.erb [hunk 2]
```diff
@@ -1,15 +1,45 @@
...
 
+<h1 style="margin-top: 130px; text-align:center;">Purchased Properties</h1>
```
Comment (ghost) [inline app/views/dashboard/_buyer.html.erb]: Per page only 1 H1 is allowed.

File: app/views/dashboard/_buyer.html.erb [hunk 3]
```diff
@@ -1,15 +1,45 @@
...
-            <%= link_to property, class: "link-a" do  %>
-              <%= render property %>
-            <% end %>
-          <% end %>
-        </div>
-      </div>
+  <div class="container">
+    <div class="row">
+      <% props.each do |property| %>
+        <% seller = User.find(property.user_id) %>
```
Comment (ghost) [inline app/views/dashboard/_buyer.html.erb]: Remove this line and move on to the controller.

File: app/views/dashboard/_buyer.html.erb [hunk 4]
```diff
@@ -1,15 +1,45 @@
...
+            <div class="col-sm-12 col-md-6" style="float:right;" >
+              <div class="testimonial-img">
+                <%= link_to property, class: "link-a" do  %>
+                  <%= render property %>
+                <% end %>
+              </div>
+            </div>
+            <div class="col-sm-12 col-md-6" style="float:left;">
+              <div class="testimonial-author-box text-center">
+                <h2>Property Details:</h2>
```
Comment (ghost) [inline app/views/dashboard/_buyer.html.erb]: manage these PROPERTY-DETAILS and SELLER-DETAILS into a separate page.

File: app/views/dashboard/_buyer.html.erb [hunk 5]
```diff
@@ -1,15 +1,45 @@
...
+              </div>
+              <div class="content">
+                <p class="testimonial-text fs-5">
+                  <b>Name: </b><%=  property.name %><br/>
+                  <b>Description: </b><%=  property.description %><br/>
```
Comment (ghost) [inline app/views/dashboard/_buyer.html.erb]: Remove extra spacing from here.

File: app/views/dashboard/_buyer.html.erb [hunk 6]
```diff
@@ -1,15 +1,45 @@
...
+                  <b>Location: </b><%=  property.address %>,<%=  property.city %><br/>
+                </p>
+              </div>
+              <div class="testimonial-author-box text-center">
+                <h2>Seller Details:</h2>
+              </div>
+              <div class="content">
+                <p class="testimonial-text fs-5">
+                  <b>Name: </b><%=  seller.name %><br/><br/><br/>
```
Comment (ghost) [inline app/views/dashboard/_buyer.html.erb]: Manage this 3 br using css styling.

File: app/views/dashboard/_seller.html.erb [hunk 7]
```diff
@@ -1,9 +1,6 @@
 <% props = Property.where(user_id: current_user.id) %>
-<% sold =  props.where(is_paid: true) %>
-<% sum = 0%>  
-<% sold.each do |sld| %>
-  <% sum = sum + sld.price %>
-<% end %>
+<% sold_property =  props.where(is_paid: true) %>
```
Comment (ghost) [inline app/views/dashboard/_seller.html.erb]: same as above.

File: app/views/dashboard/_seller.html.erb [hunk 8]
```diff
@@ -1,9 +1,6 @@
...
+
 <div class="row" style="margin-top: 130px; justify-content: center;">
```
Comment (ghost) [inline app/views/dashboard/_seller.html.erb]: Remove inline styling.

File: app/views/dashboard/sold.html.erb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+<% sold_property =  props.where(is_paid: true) %>
+
+<h1 style="text-align:center; margin-top:130px;">Sold Properties</h1>
+
+<section class="property-grid grid">
+  <div class="container">
+    <div class="row">
+      <% sold_property.each do |property| %>
+        <% buyer = User.find(property.buyer_id) %>
+        <div class="box">
```
Comment (ghost) [inline app/views/dashboard/sold.html.erb:7]: use separate partial for this content.

File: app/views/dashboard/sold.html.erb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+              </div>
+            </div>
+          </div>
+        </div>            
+      <% end %>
+    </div>
+  </div>
+</section>
+
+
```
Comment (ghost) [inline app/views/dashboard/sold.html.erb]: remove extra spacing.

File: app/views/dashboard/sold.html.erb [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+              sed aute irure.
+            </p>
+          </div>
+          <div class="w-footer-a">
+            <ul class="list-unstyled">
+              <li class="color-a">
+                <span class="color-text-a">Phone .</span> +91 12345 67890
+              </li>
+              <li class="color-a">
+                <span class="color-text-a">Email .</span> <%= mail_to "contact@propex.com", "contact@propex.com", target: '_blank' %>
```
Comment (ghost) [inline app/views/dashboard/sold.html.erb]: Can we make it like...
 
 <span class="color-text-a">Email .</span>
  <%= mail_to "contact@propex.com", "contact@propex.com", target: '_blank' %>

File: app/views/devise/registrations/edit.html.erb [hunk 12]
```diff
@@ -1,4 +1,4 @@
-<div class="text-center"><h2>Edit <%= resource_name.to_s.humanize %></h2>
+<br/><div class="text-center"><h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Comment (ghost) [inline app/views/devise/registrations/edit.html.erb]: Please try to manage using css.

File: app/views/layouts/nav/_public.html.erb [hunk 13]
```diff
@@ -6,7 +6,7 @@
         <span></span>
         <span></span>
       </button>
-      <a class="navbar-brand text-brand" href="/"><h1><b>Prop<span class="color-b">Ex</span></b></h1></a>
```
Comment (ghost) [inline app/views/layouts/nav/_public.html.erb:9]: Remove h1

File: app/views/properties/index.html.erb [hunk 14]
```diff
@@ -29,7 +29,7 @@
   <%= form_tag(search_path, :method => "get") do %>
     <span><%= text_field_tag :search, params[:search], class: "search form-control rounded", border_radius: "50px",placeholder: "Search..." ,style: "border: solid 1px #5dd95d;" %></span>
   <% end %>
-</div>
+</div><br/>
     <!-- ======= Property Grid ======= -->
     <section class="property-grid grid">
```
Comment (ghost) [inline app/views/properties/index.html.erb]: correct the indentation.

File: app/views/properties/new.html.erb [hunk 15]
```diff
@@ -5,6 +5,39 @@
...
+    <span title="Back" ><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 24 24" height="3em" width="3em" xmlns="http://www.w3.org/2000/svg"><path d="M21 11L6.414 11 11.707 5.707 10.293 4.293 2.586 12 10.293 19.707 11.707 18.293 6.414 13 21 13z"></path></svg></span>
+  <% end %>
 </div>
-</div>
\ No newline at end of file
+</div>
+
+<%# ---- Footer ---- %>
+<section class="section-footer">
+  <div class="container">
```
Comment (ghost) [inline app/views/properties/new.html.erb]: move this footer as a common partial, So we can **reuse** it.

File: app/views/dashboard/_buyer.html.erb (hunk 1 above)
Comment (rav-indra) [reply app/views/dashboard/_buyer.html.erb]: done

File: app/views/dashboard/_buyer.html.erb (hunk 2 above)
Comment (rav-indra) [reply app/views/dashboard/_buyer.html.erb]: - Done

File: app/views/dashboard/_buyer.html.erb (hunk 3 above)
Comment (rav-indra) [reply app/views/dashboard/_buyer.html.erb]: - Moved lines to controller

File: app/views/dashboard/_buyer.html.erb (hunk 4 above)
Comment (rav-indra) [reply app/views/dashboard/_buyer.html.erb]: - Made a partial for this

File: app/views/dashboard/_buyer.html.erb (hunk 5 above)
Comment (rav-indra) [reply app/views/dashboard/_buyer.html.erb]: - Removed extra spaces

File: app/views/dashboard/_buyer.html.erb (hunk 6 above)
Comment (rav-indra) [reply app/views/dashboard/_buyer.html.erb]: - Removed

File: app/views/dashboard/_seller.html.erb (hunk 7 above)
Comment (rav-indra) [reply app/views/dashboard/_seller.html.erb]: - Moved this code to controller

File: app/views/dashboard/_seller.html.erb (hunk 8 above)
Comment (rav-indra) [reply app/views/dashboard/_seller.html.erb]: - Removed inline styles and added it to CSS

File: app/views/dashboard/sold.html.erb (hunk 9 above)
Comment (rav-indra) [reply app/views/dashboard/sold.html.erb:7]: - Made partial

File: app/views/dashboard/sold.html.erb (hunk 10 above)
Comment (rav-indra) [reply app/views/dashboard/sold.html.erb]: - Removed extra spaces

File: app/views/dashboard/sold.html.erb (hunk 11 above)
Comment (rav-indra) [reply app/views/dashboard/sold.html.erb]: - Done

File: app/views/devise/registrations/edit.html.erb (hunk 12 above)
Comment (rav-indra) [reply app/views/devise/registrations/edit.html.erb]: - Done

File: app/views/layouts/nav/_public.html.erb (hunk 13 above)
Comment (rav-indra) [reply app/views/layouts/nav/_public.html.erb:9]: - Removed

File: app/views/properties/index.html.erb (hunk 14 above)
Comment (rav-indra) [reply app/views/properties/index.html.erb]: - Corrected all identetions

File: app/views/properties/new.html.erb (hunk 15 above)
Comment (rav-indra) [reply app/views/properties/new.html.erb]: - Made a partial

File: app/helpers/application_helper.rb [hunk 16]
```diff
@@ -26,6 +26,14 @@ def profile(role)
     end
   end
 
+  def c_user(value)
```
Comment (ghost) [inline app/helpers/application_helper.rb]: changes this name.

Comment (ghost) [reply app/helpers/application_helper.rb]: def navigation(is_present)
	return 'layouts/nav/user_navigation' if is_present 
	
	'layouts/nav/header'
end	

<%= render navigation(current_user.present?) %>

File: app/helpers/application_helper.rb [hunk 17]
```diff
@@ -16,6 +15,7 @@ def resource_class
   def devise_mapping
     @devise_mapping ||= Devise.mappings[:user]
   end
+
   def profile(role)
```
Comment (ghost) [inline app/helpers/application_helper.rb]: I think we can use this method more. So better to remove it.

File: app/views/dashboard/_admin.html.erb [hunk 18]
```diff
@@ -1,16 +1,15 @@
...
+      <strong><th scope="col">Role</th>
+      <th scope="col">Name</th>
+      <th scope="col">Email</th>
+    </tr>
   </thead>
   <tbody>
-      <% users.each do |user| %>
+    <% @users.each do |user| %>
       <tr>
         <th scope="row"><%= link_to image_tag(profile(user.role))%></th>
```
Comment (ghost) [inline app/views/dashboard/_admin.html.erb]: <th scope="row"><%= link_to image_tag("#{user.role}.png")%></th>

File: app/views/dashboard/_admin.html.erb [hunk 19]
```diff
@@ -19,6 +18,6 @@
         <%#= link_to "delete", user_registration_path(user), :method => :delete, :confirm => "You sure?" %>
         <%#= button_to "Delete", method: :delete, class: "btn btn-outline-danger" %>
```
Comment (ghost) [inline app/views/dashboard/_admin.html.erb]: Remove this commented code.

File: app/views/dashboard/_buyer.html.erb [hunk 20]
```diff
@@ -1,15 +1,36 @@
-<h1 style="margin-top: 120px; text-align:center;">Purchased Properties</h1>
-<% props = Property.where(buyer_id: current_user.id) %>
+<h1>Purchased Properties</h1>
```
Comment (ghost) [inline app/views/dashboard/_buyer.html.erb:1]: Remove this also.

File: app/views/dashboard/_buyer.html.erb (hunk 3 above)
Comment (ghost) [reply app/views/dashboard/_buyer.html.erb]: As per the discussion please remove this code over here.

**Comment by rav-indra** [comment]
- Removed extra spaces
- Removed extra h1 tags from pages
- Fixed Indentation
- Made partials where it was needed
- Moved inline style to CSS file
- Moved code from view to controller
- Made helpers to reduce repeated code and reuse it

File: app/controllers/dashboard_controller.rb [hunk 21]
```diff
@@ -1,5 +1,13 @@
 class DashboardController < ApplicationController
   def index
-    
+    @users = User.all
+    @props = Property.where(user_id: current_user.id)
```
Comment (ghost) [inline app/controllers/dashboard_controller.rb]: Change this line.

@user_properties = current_user.properties

File: app/views/dashboard/_buyer.html.erb [hunk 22]
```diff
@@ -1,15 +1,35 @@
...
+              <%= render partial: "property_details", locals: {property: property} %>
+              <%# Seller Details %>
+              <div class="testimonial-author-box text-center">
+                <h2>Seller Details:</h2>
+              </div>
+              <div class="content">
+                <p class="testimonial-text fs-5">
+                  <b>Name: </b><%= user(property.user_id).name %><br/>
+                  <b>Email: </b><%= user(property.user_id).email %>
+                </p>
```
Comment (ghost) [inline app/views/dashboard/_buyer.html.erb:27]: We can also write this code into better way.

<b>Name: </b><%= get_property_owner_name(property.user_id) %><br/>
<b>Email: </b><%= get_property_owner_email(property.user_id) %>



def property_owner(user_id)
    User.find_by(id: user_id)
end

def get_property_owner_name(user_id)
   user = property_owner(user_id) 

   user.name
end

def get_property_owner_email(user_id)
   user = property_owner(user_id) 

   user.email
end

File: app/views/dashboard/_seller.html.erb [hunk 23]
```diff
@@ -1,50 +1,47 @@
...
+
+<div class="row">
   <div class="col-xl-3 col-md-6 mb-4">
-      <div class="card border-left-success shadow h-100 py-2">
-          <div class="card-body">
-              <div class="row no-gutters align-items-center">
-                  <div class="col mr-2">
-                      <div class="text-xs font-weight-bold text-success text-uppercase mb-1 fs-5">
-                          My Properties</div>
-                      <div class="h5 mb-0 font-weight-bold text-gray-800" style="margin-top:-30px; text-align:end; font-size:xxx-large;"><%= props.count %></div>
```
Comment (ghost) [inline app/views/dashboard/_seller.html.erb:15]: Change this **props.count**

File: app/views/dashboard/_seller.html.erb [hunk 24]
```diff
@@ -1,50 +1,47 @@
...
-                  </div>
-              </div>
+    <div class="card border-left-success shadow h-100 py-2">
+      <div class="card-body">
+        <div class="row no-gutters align-items-center">
+          <div class="col mr-2">
+            <div class="text-xs font-weight-bold text-success text-uppercase mb-1 fs-5">
+              My Properties</div>
+            <div class="detail_box h5 mb-0 font-weight-bold text-black"><%= @props.count %></div>
```
Comment (ghost) [inline app/views/dashboard/_seller.html.erb]: same here

File: app/views/dashboard/index.html.erb [hunk 25]
```diff
@@ -1,38 +1,5 @@
-<% if current_user.admin? %>
-  <%= render 'admin' %>
-<% elsif current_user.seller?  %>
-  <%= render 'seller' %>
-<% elsif current_user.buyer? %>
-  <%= render 'buyer' %>
-<% end %>
-
+<div class="dashboard">
+  <%= render current_user.role %>
```
Comment (ghost) [inline app/views/dashboard/index.html.erb:2]: Prefect, good.

Also, Can you please rename this partials
for ex: 'buyer' to 'buyer_dashboard'

File: app/views/layouts/application.html.erb [hunk 26]
```diff
@@ -18,13 +19,7 @@
...
   <body>
     <header style="margin-bottom: 120px;">
-      <% if current_user.present? %>
-        <% p current_user%>
-        <%= render "layouts/nav/public" %>
-      <% else %>
-        <% p current_user%>
-        <%= render "layouts/nav/beforepublic"%>
-      <% end %>
+      <%= render navigation(current_user.present?) %>
```
Comment (ghost) [inline app/views/layouts/application.html.erb:22]: Prefect, good.

File: app/views/layouts/nav/_public.html.erb [hunk 27]
```diff
@@ -25,15 +25,10 @@
...
-              <% if current_user.admin? %> 
-                <%= link_to image_tag('admin.png') %>
-              <% elsif current_user.seller? %>
-                <%= link_to image_tag('seller.png') %>
-              <% elsif current_user.buyer? %>
-                <%= link_to image_tag('buyer.png') %>
-              <% end %></span>
+          <li class="nev nav-item dropdown">
+            <span class="nav-link" id="navbarDropdown" role="button" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+              <%= link_to image_tag("#{current_user.role}.png")%>
```
Comment (ghost) [inline app/views/layouts/nav/_public.html.erb:30]: Good.

**Comment by rav-indra** [comment]
- Commit

**Comment by rav-indra** [comment]
- All done

File: app/helpers/application_helper.rb (hunk 16 above)
Comment (rav-indra) [reply app/helpers/application_helper.rb]: - Done

File: app/views/dashboard/_buyer.html.erb (hunk 3 above)
Comment (rav-indra) [reply app/views/dashboard/_buyer.html.erb]: - Done

File: app/views/dashboard/_seller.html.erb (hunk 23 above)
Comment (rav-indra) [reply app/views/dashboard/_seller.html.erb:15]: - Done

File: app/views/dashboard/_seller.html.erb (hunk 24 above)
Comment (rav-indra) [reply app/views/dashboard/_seller.html.erb]: - Done

File: app/views/dashboard/index.html.erb (hunk 25 above)
Comment (rav-indra) [reply app/views/dashboard/index.html.erb:2]: - Didn't need

File: app/views/dashboard/_admin.html.erb (hunk 19 above)
Comment (rav-indra) [reply app/views/dashboard/_admin.html.erb]: - Done

File: app/views/dashboard/_admin.html.erb (hunk 18 above)
Comment (rav-indra) [reply app/views/dashboard/_admin.html.erb]: - Done

File: app/helpers/application_helper.rb (hunk 17 above)
Comment (rav-indra) [reply app/helpers/application_helper.rb]: - Done

**Comment by rav-indra** [comment]
- All done

**Pull request merged**